include/flux/op/set_algorithms.hpp

```diff
@@ -0,0 +1,576 @@
+                         std::equality_comparable<cursor_t<Base2>>
+                = default;
+        };
+
+        template <typename Self>
+        static inline constexpr bool maybe_const_iterable
+            = std::is_const_v<Self> ? flux::sequence<Base1 const> : true;
+
+        template <typename Self>
+        static constexpr void update(Self& self, cursor_type& cur) {
+            while(not flux::is_last(self.base1_, cur.base1_cursor) && 
+                  not flux::is_last(self.base2_, cur.base2_cursor))
+            {
+                if(std::invoke(self.cmp_, flux::read_at(self.base1_, cur.base1_cursor), 
+                                          flux::read_at(self.base2_, cur.base2_cursor))) {
+                    flux::inc(self.base1_, cur.base1_cursor);
+                } else {
+
+                    if(not std::invoke(self.cmp_, flux::read_at(self.base2_, cur.base2_cursor), 
+                                                  flux::read_at(self.base1_, cur.base1_cursor))) {
+                        return;
+                    }
+                    flux::inc(self.base2_, cur.base2_cursor);
+                }
+            }
+        }
+
+    public:
+        using value_type = value_t<Base1>;
+
+        inline static constexpr bool is_infinite = flux::infinite_sequence<Base1>;
+
+        template <typename Self>
+            requires maybe_const_iterable<Self>
+        static constexpr auto first(Self& self) -> cursor_type
+        {
+            auto cur = cursor_type{.base1_cursor = flux::first(self.base1_),
+                                   .base2_cursor = flux::first(self.base2_)};
+            update(self, cur);
+            return cur;
+        }
+
+        template <typename Self>
+            requires maybe_const_iterable<Self>
+        static constexpr auto is_last(Self& self, cursor_type const& cur) -> bool
+        {
+            return flux::is_last(self.base1_, cur.base1_cursor) || 
+                   flux::is_last(self.base2_, cur.base2_cursor);
+        }
+
+        template <typename Self>
+            requires maybe_const_iterable<Self>
+        static constexpr auto inc(Self& self, cursor_type& cur) -> void
+        {
+            flux::inc(self.base1_, cur.base1_cursor);
+            flux::inc(self.base2_, cur.base2_cursor);
+            update(self, cur);
+        }
+
+        template <typename Self>
+            requires maybe_const_iterable<Self>
+        static constexpr auto read_at(Self& self, cursor_type const& cur)
+            -> decltype(flux::read_at(self.base1_, cur.base1_cursor))
+        {
+            return flux::read_at(self.base1_, cur.base1_cursor);
+        }
+
+        template <typename Self>
+            requires maybe_const_iterable<Self> && bounded_sequence<Base1>
+        static constexpr auto last(Self& self) -> cursor_type
+        {
+            return cursor_type{flux::last(self.base1_), flux::last(self.base2_)};
+        }
```

Comment:
I think this has the same problem as for `set_difference()`: we say that iteration for `set_intersection` is complete when either underlying sequence is exhausted, but when comparing cursors for equality we require both underlying cursors to be the same... and so a cursor for which `is_last()` is true can end up being not equal to a cursor returned from `last()`, which is really weird.

So again, I think the best solution is to remove the `last()` function for `set_intersection_adaptor`.